reconcile/openshift_clusterrolebindings.py

```diff
@@ -45,119 +55,162 @@
+            "subjects": [{"kind": "User", "name": user}],
+        }
+        return OCResource(
+            resource=OR(
+                body,
+                QONTRACT_INTEGRATION,
+                QONTRACT_INTEGRATION_VERSION,
+                error_details=name,
+            ),
+            resource_name=name,
+        )
+
+    def construct_sa_oc_resource(
+        self, sa_namespace_name: str, sa_name: str
+    ) -> OCResource:
+        name = f"{self.cluster_role_name}-{sa_namespace_name}-{sa_name}"
+        body: dict[str, Any] = {
+            "apiVersion": "rbac.authorization.k8s.io/v1",
+            "kind": "ClusterRoleBinding",
+            "metadata": {"name": name},
+            "roleRef": {"name": self.cluster_role_name, "kind": "ClusterRole"},
+            "subjects": [
+                {
+                    "kind": "ServiceAccount",
+                    "name": sa_name,
+                    "namespace": sa_namespace_name,
+                }
+            ],
+        }
+        return OCResource(
+            resource=OR(
+                body,
+                QONTRACT_INTEGRATION,
+                QONTRACT_INTEGRATION_VERSION,
+                error_details=name,
+            ),
+            resource_name=name,
+        )
+
+    @staticmethod
+    def get_usernames_from_users_and_cluster(
+        users: list[UserV1], cluster: ClusterV1
+    ) -> set[str]:
+        auth_dict = [auth.model_dump(by_alias=True) for auth in cluster.auth]
+        user_keys = ob.determine_user_keys_for_access(cluster.name, auth_dict)
+        return {
+            name
+            for user in users or []
+            for user_key in user_keys or []
+            if (name := getattr(user, user_key, None))
+        }
 
 
 def fetch_desired_state(
-    ri: ResourceInventory | None, oc_map: ob.ClusterMap
-) -> list[dict[str, str]]:
-    gqlapi = gql.get_api()
-    roles: list[dict] = expiration.filter(gqlapi.query(ROLES_QUERY)["roles"])
-    users_desired_state = []
-    # set namespace to something indicative
-    namespace_cluster_scope = "cluster"
-    for role in roles:
-        permissions = [
-            {"cluster": a["cluster"], "cluster_role": a["clusterRole"]}
-            for a in role["access"] or []
-            if a["cluster"] and a["clusterRole"]
-        ]
-        if not permissions:
+    ri: ResourceInventory | None, allowed_clusters: set[str] | None = None
+) -> None:
+    if allowed_clusters is not None and not allowed_clusters:
+        return None
+    cluster_roles: list[RoleV1] = expiration.filter(get_app_interface_clusterroles())
+    clusters_to_check = allowed_clusters or set()
+    cluster_role_binding_specs = [
+        cluster_role_binding_spec
+        for cluster_role in cluster_roles
+        for cluster_role_binding_spec in ClusterRoleBindingSpec.create_cluster_role_binding_specs(
+            cluster_role
+        )
+        if cluster_role_binding_spec.cluster.name in clusters_to_check
+    ]
```

Comment:
when `allowed_clusters` is `None`, means no filter, `cluster_role_binding_specs` here will be empty, expected?